### PR TITLE
reduce verbosity of `reHLT` customisation in `SKIM` step

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR is a follow-up of #40481, addressing https://github.com/cms-sw/cmssw/pull/40481#issuecomment-1400550727.

It reduces the verbosity of `cmsDriver.py` when the `re-HLT` customisation is applied to the SKIM step. In this case, it produces only a 1-line message, instead of 1 line for every affected Sequence of the SKIM step.

Merely technical. No changes expected, except for a reduction of printouts from `cmsDriver.py`.

#### PR validation:

Some basic manual checks with `runTheMatrix.py` and `cmsDriver.py`.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

N/A

### CODE REVIEW

File: Configuration/Applications/python/ConfigBuilder.py [hunk 1]
```diff
@@ -1942,17 +1949,20 @@ def renameInputTagsInSequence(self,sequence,oldT="rawDataCollector",newT="rawDat
...
+        if verbosityLevel > 0:
+            print("replacing %s process name - sequence %s will use '%s'" % (HLTprocess, sequence, proc))
+        verboseVisit = (verbosityLevel > 1)
+        getattr(self.process,sequence).visit(ConfigBuilder.MassSearchReplaceProcessNameVisitor(HLTprocess, proc, whitelist = ("subSystemFolder",), verbose = verboseVisit))
         if 'from Configuration.Applications.ConfigBuilder import ConfigBuilder' not in self.additionalCommands:
             self.additionalCommands.append('from Configuration.Applications.ConfigBuilder import ConfigBuilder')
-        self.additionalCommands.append('process.%s.visit(ConfigBuilder.MassSearchReplaceProcessNameVisitor("%s", "%s", whitelist = ("subSystemFolder",)))'% (sequence,HLTprocess, proc))
-
+        self.additionalCommands.append(
+          'process.%s.visit(ConfigBuilder.MassSearchReplaceProcessNameVisitor("%s", "%s", whitelist = ("subSystemFolder",), verbose = %s))'% (sequence, HLTprocess, proc, verboseVisit))
```
Comment (missirol) [inline Configuration/Applications/python/ConfigBuilder.py:1965]: This adds a bit of text to (some of) the cfgs produced by `cmsDriver.py`. Specifically, it adds a `, verbose = False` in the reHLT-related customisations inside the cfg.

**Comment by missirol** [comment]
please test

**Comment by perrotta** [comment]
test parameters:
- workflows = 4.17

**Comment by perrotta** [comment]
please test

**Comment by missirol** [comment]
There is something I don't understand about the logs from `runTheMatrix.py`. Take wf `136.731`, step 3:

 1. [latest tests of this PR](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-208f86/30161/runTheMatrix-results/136.731_RunSinglePh2016B/step3_RunSinglePh2016B.log)
 2. [baseline (from IB)](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_13_0_X_2023-01-24-2300/el8_amd64_gcc11/-GenuineIntel/matrix-results/136.731_RunSinglePh2016B/step3_RunSinglePh2016B.log)
 3. [latest tests of another, unrelated, PR](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e8d35a/30181/runTheMatrix-results/136.731_RunSinglePh2016B/step3_RunSinglePh2016B.log)

I don't understand why 1. does not include printouts from `cmsDriver.py` like `Step: RECO Spec:`. It might be due to this PR, but I do not see how.

**Comment by missirol** [comment]
> No significant changes to the logs found

@smuzaffar, I have a couple of questions. In this PR we expect a reduction in the logs of the matrix tests.

 - Does the bot only report increases in the logs, or also reductions? [(here?)](https://github.com/cms-sw/cms-bot/blob/e744d945b54abd134497d5b276df8fcab0dcef4c/logRootQA.py#L339-L342)

 - In the outputs of PR tests (e.g. [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-208f86/30161/)), can one find some 'global' stats on the logs? (total number of lines, or similar)

**Comment by missirol** [comment]
unhold

I think this PR is okay (the last push is only a rebase).

What was observed in https://github.com/cms-sw/cmssw/pull/40599#issuecomment-1405223039 is, I think, due to the fact that buffers to stdout/err should be [flushed before calling `os.execvpe`](https://docs.python.org/3/library/os.html#os.execvpe) (another couple of random links [[1]](https://comp.unix.programmer.narkive.com/43SJNl1P/what-happens-to-stdout-stderr-after-execvp#post3) [[2]](https://adamj.eu/tech/2020/06/26/how-to-check-if-pythons-output-buffering-is-enabled/)).

> I don't understand why 1. does not include printouts from `cmsDriver.py` like `Step: RECO Spec:`. It might be due to this PR, but I do not see how.

This PR reduces the printouts of certain steps to the point where the buffer size is not exceeded, so those printouts are then lost when `os.execvpe` is called (because the buffers are not flushed beforehand).

The lack of some of the outputs of `cmsDriver.py` (e.g. lines like "Step: XX Spec:") in the logs of the matrix tests can be seen in recent tests of unrelated PRs (e.g. [[1]](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5c0c32/30223/runTheMatrix-results/11634.0_TTbar_14TeV+2021/step2_TTbar_14TeV+2021.log) [[2]](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5c0c32/30223/runTheMatrix-results/11634.0_TTbar_14TeV+2021/step3_TTbar_14TeV+2021.log) [[3]](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5c0c32/30223/runTheMatrix-results/11634.0_TTbar_14TeV+2021/step4_TTbar_14TeV+2021.log) [[4]](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5c0c32/30223/runTheMatrix-results/11634.0_TTbar_14TeV+2021/step5_TTbar_14TeV+2021.log) from wf `11634.0`). For reasons unknown to me, this does not happen with the matrix outputs of the IB baseline (e.g. [[1]](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_13_0_X_2023-01-28-1100/el8_amd64_gcc11/-GenuineIntel/matrix-results/11634.0_TTbar_14TeV+2021/step2_TTbar_14TeV+2021.log) [[2]](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_13_0_X_2023-01-28-1100/el8_amd64_gcc11/-GenuineIntel/matrix-results/11634.0_TTbar_14TeV+2021/step3_TTbar_14TeV+2021.log) [[3]](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_13_0_X_2023-01-28-1100/el8_amd64_gcc11/-GenuineIntel/matrix-results/11634.0_TTbar_14TeV+2021/step4_TTbar_14TeV+2021.log) [[4]](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_13_0_X_2023-01-28-1100/el8_amd64_gcc11/-GenuineIntel/matrix-results/11634.0_TTbar_14TeV+2021/step5_TTbar_14TeV+2021.log) from wf `11634.0`).

#40635 adds the call to flush the stdout buffer before `os.execvpe` is used.

**Comment by missirol** [comment]
test parameters:
  - workflows = 4.17
  - pull_requests = #40635

**Comment by missirol** [comment]
please test

This PR does not require #40635 (I test them together to properly see how this PR modifies the printouts of `cmsDriver.py`).

**Comment by missirol** [comment]
The RelVals-INPUT error is unrelated (the tests passed in https://github.com/cms-sw/cmssw/pull/40599#issuecomment-1402868163, and there no real changes since then).

> No significant changes to the logs found

As noted in https://github.com/cms-sw/cmssw/pull/40599#issuecomment-1405225103, I'm not sure the bot reports reductions in the logs.

One can now see the reduction in the logs due to this PR, e.g. [this](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-208f86/30231/runTheMatrix-results/136.731_RunSinglePh2016B/step3_RunSinglePh2016B.log) vs [baseline](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_13_0_X_2023-01-28-1100/el8_amd64_gcc11/-GenuineIntel/matrix-results/136.731_RunSinglePh2016B/step3_RunSinglePh2016B.log).

**Comment by perrotta** [comment]
+1

**Comment by perrotta** [comment]
merge

**Comment by smuzaffar** [comment]
> * Does the bot only report increases in the logs, or also reductions? [(here?)](https://github.com/cms-sw/cms-bot/blob/e744d945b54abd134497d5b276df8fcab0dcef4c/logRootQA.py#L339-L342)

@missirol , yes bot was only reporting the increase in logs output but I now have updated it to report any reduction too. From the [comparison tests log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-28-1100+208f86/55215/validateJR/logRootQA-events.log) I see that 1286 log lines were removed.
```
Logs: -1286 False 216 299 0
Events: True 166 0.0 49 [] []
SUMMARY No significant changes to the logs found
```

> * In the outputs of PR tests (e.g. [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-208f86/30161/)), can one find some 'global' stats on the logs? (total number of lines, or similar)

you can find some information by going to [Comparison with the baseline)(https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-28-1100+208f86/55215/) and then look into [validateJR/logRootQA-events.log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-28-1100+208f86/55215/validateJR/logRootQA-events.log) file

**Comment by missirol** [comment]
great, thanks @smuzaffar !

**Comment by smuzaffar** [comment]
> There is something I don't understand about the logs from `runTheMatrix.py`. Take wf `136.731`, step 3:
> 
> 1. [latest tests of this PR](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-208f86/30161/runTheMatrix-results/136.731_RunSinglePh2016B/step3_RunSinglePh2016B.log)
> 2. [baseline (from IB)](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_13_0_X_2023-01-24-2300/el8_amd64_gcc11/-GenuineIntel/matrix-results/136.731_RunSinglePh2016B/step3_RunSinglePh2016B.log)
> 3. [latest tests of another, unrelated, PR](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e8d35a/30181/runTheMatrix-results/136.731_RunSinglePh2016B/step3_RunSinglePh2016B.log)
> 
> I don't understand why 1. does not include printouts from `cmsDriver.py` like `Step: RECO Spec:`. It might be due to this PR, but I do not see how.

I was also looking in to it and now after reading this comment I think the problem is that bot might have set ` PYTHONUNBUFFERED` only for IB tests. This explains why we see extra printouts in IB logs and not in PR logs. I will update bot to set ` PYTHONUNBUFFERED` for PR too.

**Comment by missirol** [comment]
Thanks for having a look. Indeed, that's the only explaination I could think of, but I could not see where `PYTHONUNBUFFERED` is set for the IB tests. Out of curiosity, where is that done?

**Comment by smuzaffar** [comment]
all IB tests use https://github.com/cms-sw/cms-bot/blob/master/run-ib-testbase.sh to create dev area and this scripts sets `export PYTHONUNBUFFERED=1`

**Comment by smuzaffar** [comment]
by the way, bot pr https://github.com/cms-sw/cms-bot/pull/1927 should fix this for PR tests too

**Comment by missirol** [comment]
Sounds good. Thanks for the info!

**Pull request merged**